lib/radiator/directory/editor.ex

```diff
@@ -458,4 +458,89 @@ defmodule Radiator.Directory.Editor do
       @not_authorized_match
     end
   end
+
+  ## User Permission Management
+
+  # TODO: list all collaborators of underlying entities as well.
+
+  @spec list_collaborators(Auth.User.t(), Network.t()) ::
+          [
+            Collaborator.t()
+          ]
+          | {:error, any}
+  def list_collaborators(actor = %Auth.User{}, subject = %Network{}) do
+    if has_permission(actor, subject, :manage) do
+      query =
+        from p in Ecto.assoc(subject, :permissions),
+          join: u in Auth.User,
+          on: p.user_id == u.id,
+          join: s in Network,
+          on: s.id == p.subject_id,
+          preload: [user: u]
+
+      with perm_list when is_list(perm_list) <- Repo.all(query) do
+        perm_list
+        |> Enum.map(fn
+          perm = %Radiator.Perm.Permission{} ->
+            %Collaborator{user: perm.user, permission: perm.permission, subject: subject}
+        end)
+        |> Enum.sort(fn
+          a, b ->
+            case Radiator.Perm.Ecto.PermissionType.compare(a.permission, b.permission) do
+              :gt -> true
+              :eq -> a.user.name < b.user.name
+              :lt -> false
+            end
+        end)
+      end
+    else
+      @not_authorized_match
+    end
+  end
+
+  # TODO: Handle the accidental update case
+  @spec add_collaborator(Auth.User.t(), Collaborator.t()) ::
+          {:error, any} | {:ok, Collaborator.t()}
+  def add_collaborator(
+        actor = %Auth.User{},
+        collaborator = %Collaborator{user: user, subject: subject, permission: permission}
+      )
+      when is_permission(permission) do
+    if has_permission(actor, subject, :manage) do
+      case Editor.Permission.set_permission(user, subject, permission) do
```

Comment:
Use `with` to avoid `other -> other`? (applies to all 3 `add_collaborator` functions). 

```elixir
with :ok <- Editor.Permission.set_permission(user, subject, permission), do: {:ok, collaborator}
```

Reply:
ok, can do. was a little ambigous of that because 
* it is nested already in an if (which we could also move to a more general expression used with a with?)
* i might want to handle some errors and then i'll end up back in this land. `other -> other` seems to be a bad pattern in general though. maybe something more suitably named? Also the internal submodules of Editor should probably not return something we passthrough directly anyways, so a match against a changeset we want to bubble up and other errors might be more appropriate here anyways

Reply:
There is a pattern I have seen using `with` and "tagged" conditions that is a bit verbose but linearizes otherwise nested conditional if/case/with constructs:

```elixir
with {:permission, true} <- {:permission, has_permission(actor, subject, :manage)},
     {:set, :ok} <- {:set, Editor.Permission.set_permission(user, subject, permission)} do
  {:ok, collaborator}
else
  {:permission, _} -> @not_authorized_match
  {:set, _} -> raise {:error, :could_not_set_permission}
end
```

You code the "happy path" in the with but tag each line with a unique tuple. Then you can check in the `else` match which line failed and return the error accordingly.

Reply:
ack